Blockfrost chain integration

v0d1ch opened the pull request:
- Final step for the blockfrost integration.

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

v0d1ch committed:
Put Wallet types into separate module

Also introduce Hydra.Chain.Blockfrost.Wallet

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Put Wallet types into separate module

Also introduce Hydra.Chain.Blockfrost.Wallet

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Put Wallet types into separate module

Also introduce Hydra.Chain.Blockfrost.Wallet

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Introduce queryWalletInfo and new module Hydra.Chain.Blockfrost

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Simplify blockfrost setup

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Simplify blockfrost setup

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Simplify blockfrost setup

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Simplify blockfrost setup

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Update changelog

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Construct a TimeHandle using blockfrost

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Construct a TimeHandle using blockfrost

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Construct a TimeHandle using blockfrost

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

ch1bo commented on hydra-cardano-api/src/Cardano/Api/UTxO.hs:
```diff
@@ -76,6 +76,9 @@ filter fn = UTxO . Map.filter fn . toMap
 inputSet = Map.keysSet . toMap
 
+squash :: [UTxO' out] -> UTxO' out
+squash = foldMap (<> mempty)
```
Must: use `fold` and not introduce this function

ch1bo commented on hydra-node/src/Hydra/Options.hs:
```diff
@@ -407,6 +407,8 @@ data BlockfrostChainConfig = BlockfrostChainConfig
   , cardanoSigningKey :: FilePath
   -- ^ Path to the cardano signing key of the internal wallet.
+  , hydraScriptsTxId :: [TxId]
+  -- ^ Identifier of transaction holding the hydra scripts to use.
```
Must: not replicate command line options

I think I demonstrated two weeks ago, that command line options like `--cardano-signing-key` and `--hydra-scripts-tx-id` are common to both, `Direct` and `Blockfrost` chain backends. Hence, we should keep them not in `Direct` or `Blockfrost` specific data structures. This will directly drive function structure and code re-use.

v0d1ch committed:
Remove squash in favor of fold

fold already has a Monoid and Foldable constraints so it is convenient.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Draft common types between BF/Cardano

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

ch1bo committed:
Rename DirectBackend fields

ch1bo committed:
Draft usage of ChainBackend

v0d1ch committed:
Unify blockfrost qurying functions to match cardano ones

Being consistent with the parameters and naming should help out.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Start using CardanoChainConfig

Still keeping around DirectChainConfig until we decide how to
abstract away the details (typeclass, handle pattern, etc.)

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Add class for abstract operations over ChainBackend

Implement queryScriptRegistry and queryGenesisParameters
to see how it looks in action.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Add queryNetworkId and loadChainContext using it

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Add more queries to make mkTinyWallet work with CardanoChainConfig

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Add queryTimeHandle to BackendOps and start withDirectChain port

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Use generic-lens to disambiguate fields in OptionsSpec

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Add blockfrost backend for chain following

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Connect CardanoChainConfig in Options

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Remove Direct option from hydra-node

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Compile hydra-cluster

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Properly convert to cardano value

We were not handling correctly PolicyId/AssetName conversions and
we were also disregarding any datums/scripts.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Draft a test for blockfrost integration

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Remap needed test functions to work with blockfrost

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Improve Blocfrost.Client to run in BlockfrostClientT

Add await and run all blockfrost functions in one do block to reduce
posibility of things going wrong but still no luck.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

ch1bo commented:
Please update title and description to reflect that this is now the main blockfrost chain backend integration PR.

This looks already quite good! You asked for potential further refactoring of the way you switch between backends. Currently this is a bit weird using a type class with only one instance and then switching on the sum of that type all the time. I see various cleaner options:

1. Keep `ChainBackend` and make all backend ops just top level functions.
    - This keeps direct and blockfrost functions in Hydra.Node (not nice)
1. Make `ChainBackend` a type class and create two instances, e.g. for existing option types `DirectBackend` and `BlockfrostBackend`
    - This separates at least direct and blockfrost logic into their respective module
1. Follow our component design (ADR3 or so) where the current `BackendOps` are moved into a "down" handle that we just call `ChainBackend` and the `ChainSyncHandler` becoming the "up" callback part.
    - this is a bit more involved but consistent with the remaining application architecture

v0d1ch committed:
Fix waiting logic for blockfrost

Added function to wait for specific UTxO. I'd like to refactor
blockfrost client and keep only the least amount of code possible.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Reverse steps and memory in pparams

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Use CostModelsRaw to construct plutus cost models

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Add code to actually post txs from Direct chain blockfrost component

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Faucet needs to wait for _user_ utxo too

Also a small refactor in the faucet.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Add timeout to queryUTxO/queryUTxOByTxIn

Make queryUTxOByTxIn take in tx hash as Text
Increase utxo amounts in test to cover for script execution.
Small refactors here and there

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Fix the blockfrost observation

Problem was that we were starting from scratch each time because of
retry. Using TMVar to record block hash solved this problem.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Add proper metadata to abort tx

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Improve the typeclass design

This slightly improves the design by adding BackendOps
instances for underlying types and then have one instance for
ChainBackend. Too bad we can't derive this instance (yet).

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Make all packages compile again

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Expose submitTransaction for blockfrost

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Alter changelog, remove deleted logs.yaml...

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Constrain blockfrost-client dependency to hydra-node

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Remove generic-lens dependency from hydra-cluster

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Rebase changes

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

ch1bo commented on CHANGELOG.md:
````diff
@@ -41,6 +41,8 @@ changes.
     ETCD_AUTO_COMPACTION_RETENTION=168h
     ```
 
+- Enable blockfrost integration for hydra-node.
````
This is in the wrong place (but changelog is shuffled here too, see my comment in #1945)

ch1bo commented on hydra-node/src/Hydra/Node/Run.hs:
```diff
@@ -151,9 +151,8 @@ getGlobalsForChain = \case
-    queryGenesisParameters networkId nodeSocket QueryTip
-      >>= newGlobals
+  Cardano CardanoChainConfig{chainBackend} ->
+    queryGenesisParameters chainBackend >>= newGlobals
```
:+1:

ch1bo commented on hydra-node/test/Hydra/OptionsSpec.hs:
Must: add a test that `hydra-node --blockfrost <path>` results in a `BlockfrostBackend` in `Options`

ch1bo commented on hydra-node/src/Hydra/Options.hs:
```diff
@@ -137,32 +137,50 @@ data PublishOptions = PublishOptions
+defaultBlockfrostBackend =
+  BlockfrostBackend
+    { projectPath = "blockfrost-project.txt"
     }
```
Should: not need a `defaultBlockfrostBackend`

We want to go away from defaults (#1064) and this is only used in the arbitrary instance. Just move it there, then it also does not matter whether this is a commonly used name (e.g. in blockfrost docs).

ch1bo commented on hydra-node/src/Hydra/Options.hs:
```diff
@@ -555,7 +570,7 @@ nodeSocketParser =
     ( long "node-socket"
         <> metavar "FILE"
-        <> value defaultDirectChainConfig.nodeSocket
+        <> value "node.socket"
```
Should: not hard-code this default here if we have a `defaultDirectBackend.nodeSocket`

ch1bo commented on weeder.toml:
```diff
@@ -12,10 +12,11 @@ roots = [
  , "toCardanoGenesisParameters"
  , "mkTinyWallet"
  , "queryTimeHandle"
+ , "loadChainContext"
```
Must: not need these exceptions anymore (as this is now the full integration)

ch1bo commented on hydra-node/hydra-node.cabal:
```diff
@@ -329,6 +331,7 @@ test-suite tests
     , contra-tracer
     , directory
     , filepath
+    , generic-lens
```
Why do we need this dependency?

Could this be done in a separate refactoring PR?

v0d1ch replied:
We need it to solve ambiguous record updates

ch1bo commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -582,10 +582,15 @@ spec = around (showLogsOnFailure "EndToEndSpec") $ do
+                            DirectBackend
+                              { networkId = Hydra.Options.networkId defaultDirectBackend
+                              , nodeSocket = nodeSocket
+                              }
```
```suggestion
                            defaultDirectBackend { nodeSocket }
```

v0d1ch replied:
We can't do this since we get 
```
The record update defaultCardanoChainConfig
                        {chainBackend = Blockfrost
                                          (BlockfrostBackend
                                             "blockfrost-project.txt")} with type CardanoChainConfig is ambiguous.
    This will not be supported by -XDuplicateRecordFields in future releases of GHC.
    |
336 |             { chainConfig = Cardano (defaultCardanoChainConfig {chainBackend = Blockfrost (BlockfrostBackend "blockfrost-project.txt")})

```
This is why I needed generic-lens package in another comment you made.

ch1bo commented on hydra-cluster/test/Test/DirectChainSpec.hs:
```diff
@@ -375,6 +379,88 @@ spec = around (showLogsOnFailure "DirectChainSpec") $ do
             failAfter 5 $
               waitForUTxO node expectedUTxO
 
+  it "can open, close & fanout a Head using Blockfrost" $ \tracer -> do
```
The naming in this PR is very confusing.

Option-wise we have `Direct` and `Blockfrost`, but this test and all your changes to `withDirectChain` suggest that the `Hydra.Chain.Direct` module is used by both.. the direct chain connection and the blockfrost connection.

v0d1ch replied:
Moved the test to separate module related to Blockfrost.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct.hs:
```diff
@@ -238,14 +226,10 @@ withDirectChain tracer config ctx wallet chainStateHistory callback action = do
+newtype ConnectException = ConnectException
   { ioException :: IOException
-  , nodeSocket :: SocketPath
-  , networkId :: NetworkId
```
Should: not remove this information.

This comes from the fact that you are generalizing this module to be for both, direct chain connection and blockfrost. However, there is still direct chain specifics like this exception handling (rethrowing with more context).

ch1bo commented on hydra-node/src/Hydra/Chain/Direct.hs:
```diff
@@ -165,7 +148,7 @@ mkTinyWallet tracer config = do
 withDirectChain ::
   Tracer IO DirectChainLog ->
-  DirectChainConfig ->
+  CardanoChainConfig ->
```
I like that you could re-use lots of this function and would suggest this ought to be named `withCardanoChain :: CardanoChainConfig -> ... -> ChainComponent Tx IO a` Consequently we should be moving the generic pieces to Hydra.Chain.Cardano while the actual connection to the node (the "direct" parts) should remain here.

ch1bo commented on hydra-node/src/Hydra/Node.hs:
```diff
@@ -420,3 +445,119 @@ data HydraNodeLog tx
 deriving stock instance IsChainState tx => Show (HydraNodeLog tx)
 deriving anyclass instance IsChainState tx => ToJSON (HydraNodeLog tx)
+
+class BackendOps a where
```
IMO this is the wrong module for this. It is switching between `Hydra.Chain.Blockfrost` and `Hydra.Chain.Direct`, both being "chain backends".

I would suggest to:
- Create a new module `Hydra.Chain.Backend` which contains this class, renamed to `class ChainBackend a`
- Rename the types in `Options` to `DirectOptions` and `BlockfrostOptions` to avoid conflicts
- Create a type and instance of `ChainBackend` in each `Hydra.Chain.Direct` and `Hydra.Chain.Blockfrost`, for example:
```haskell
data DirectBackend = DirectBackend { options :: DirectOptions }

instance ChainBackend DirectBackend where
```
  - This is primarily to avoid orphan instances
- Each backend then also has a `mkDirectBackend :: DirectOptions -> DirectBackend` or `withDirectBackend :: DirectOptions -> (DirectBackend -> m a) -> m a` depending on whether we want to initialize some state.

Alternative:
- Use handles and skip the orphan instance avoidance, i.e.
```haskell
data ChainBackend m = ChainBackend { queryTip :: m ChainPoint, ... }

mkDirectChainBackend :: m (ChainBackend m)
mkBlockfrostChainBackend :: m (ChainBackend m)
```

ch1bo commented on hydra-node/src/Hydra/Chain/Blockfrost/TimeHandle.hs:
Why is this not also part of `Blockfrost.Client` like `queryProtocolParameters`?

ch1bo replied:
Looking at it again.. `queryTimeHandle` should even be defined in terms of `ChainBackend` as nothing of it is specific to `Blockfrost` and we can move it to a `Hydra.Chain.TimeHandle` (with all other `TimeHandle` things)`

ch1bo commented on hydra-node/src/Hydra/Chain/Blockfrost/Client.hs:
```diff
@@ -63,153 +83,83 @@ runBlockfrostM prj action = do
+            let txin = toCardanoTxIn unTxHash _addressUtxoOutputIndex
+             in toCardanoUTxO networkId txin _addressUtxoAddress _addressUtxoReferenceScriptHash _addressUtxoDataHash _addressUtxoAmount _addressUtxoInlineDatum
+      )
+      addressUTxO
```
Isn't this the same as `queryUTxO`?

v0d1ch replied:
Apparently it is :) I used queryUTxO directly, thanks

ch1bo commented on hydra-node/src/Hydra/Chain/Blockfrost/Client.hs:
```diff
@@ -63,153 +83,83 @@ runBlockfrostM prj action = do
-    Left err -> error (show err)
-    Right p -> p
+  BlockfrostClientT IO [TxId]
+publishHydraScripts sk = do
```
Should: As we have a nice interface to the specifics of accessing the cardano chain (currently a type class), we should be able to write a function `publishHydraScripts :: ChainBackend b => b -> SigningKey PaymentKey -> IO [TxId]`

ch1bo commented on hydra-node/src/Hydra/Chain/Blockfrost/Client.hs:
```diff
@@ -277,22 +227,167 @@ toCardanoPParams = do
+  , drepDeposit :: Integer
+  , drepActivity :: Blockfrost.Quantity
+  , minFeeRefScriptCostPerByte :: NonNegativeInterval
+  }
```
Where is this used?

v0d1ch replied:
Only in this module, helps in keeping track of all these fields more easily.

ch1bo commented on hydra-node/src/Hydra/Chain/Blockfrost/Client.hs:
```diff
@@ -277,22 +227,167 @@ toCardanoPParams = do
+toCardanoNetworkId :: Integer -> NetworkId
+toCardanoNetworkId = \case
+  0 -> Mainnet
+  magicNbr -> Testnet (NetworkMagic (fromInteger magicNbr))
```
Must: map mainnet network magic to `Mainnet` like `fromNetworkMagic` from `cardano-api` does

v0d1ch replied:
We should do this also in hydra-chain-observer

ch1bo commented on hydra-node/src/Hydra/Chain/Blockfrost/Client.hs:
```diff
@@ -277,22 +227,167 @@ toCardanoPParams = do
+addressToText :: AddressInEra -> Text
+addressToText = \case
+  ShelleyAddressInEra addr -> serialiseToBech32 addr
+  ByronAddressInEra{} -> error "Byron."
```
Should: use `serialiseAddress` from cardano api instead

ch1bo commented on hydra-node/src/Hydra/Chain/Blockfrost/Client.hs:
```diff
@@ -277,22 +227,167 @@ toCardanoPParams = do
+textAddrOf networkId vk = addressToText (mkVkAddress @Era networkId vk)
+
+toTxHash :: TxId -> Text
+toTxHash (TxId hash) = hashToTextAsHex hash
```
Should: use `serialiseToRawBytesHexText` from cardano api instead

ch1bo commented on hydra-node/src/Hydra/Chain/Blockfrost.hs:
```diff
@@ -2,61 +2,114 @@ module Hydra.Chain.Blockfrost where
   Tracer IO DirectChainLog ->
   BlockfrostChainConfig ->
   IO (TinyWallet IO)
 mkTinyWallet tracer config = do
```
Nothing of this is blockfrost specific (as far as I can see) and we should be able to define this for any `ChainBackend`.

ch1bo commented on hydra-node/src/Hydra/Chain/Blockfrost.hs:
```diff
@@ -2,61 +2,114 @@ module Hydra.Chain.Blockfrost where
+  -- | Hydra party of our hydra node.
+  Party ->
+  IO ChainContext
+loadChainContext config party = do
```
Also this.. nothing blockfrost specific (assuming we have a `queryNetworkId :: m NetworkId`) and can be defined in terms of `ChainBackend`

ch1bo commented on hydra-node/src/Hydra/Chain/Blockfrost.hs:
```diff
@@ -65,3 +118,175 @@ mkTinyWallet tracer config = do
+blockfrostChain ::
+  (MonadIO m, MonadCatch m, MonadAsync m, MonadDelay m) =>
+  Tracer m DirectChainLog ->
+  TQueue m (Tx, TMVar m (Maybe (PostTxError Tx))) ->
```
This is taken a bit out of context (from `Hydra.Chain.Direct`) and will make hard to understand code. Instead we should aim for an abstraction where the blockfrost chain backend has a means to wait for next tx to submit (and return success or an error). The fact that this is a queue should not leak all over the place.

v0d1ch replied:
I don't quite follow you on this one...anyway, now we have `withBlockfrostChain` so everything is encapsulated in there.

ch1bo commented on hydra-node/src/Hydra/Chain/Blockfrost.hs:
```diff
@@ -65,3 +118,175 @@ mkTinyWallet tracer config = do
+  current <- readTVarIO stateTVar
+  nextHash <- rollForward tracer prj handler wallet blockConfirmations current
+  atomically $ writeTVar stateTVar nextHash
+  loop tracer prj handler wallet blockConfirmations stateTVar
```
This is quite contrived usage of `TVar`s.. also the signature of this function is longer than its definition. How about moving this into `blockfrostChainFollow` and shape also use the next hash as an argument. Something like:
```haskell
follow tracer prj handler wallet = do
  hash <- blockHash <$> queryTip
  loop hash
 where
  blockConfirmations = 1

  loop current = do
    nextHash <- rollForward current
    loop nextHash

  rollForward blockHash = ...
```

No TVar needed and WAY less arguments. Also, it is very explicit that `blockConfirmations` is a constant.

v0d1ch replied:
I agree, this can be simplified but the usage of TVar is the easiest thing to do here since we need to keep somehow the state of current block hash across `retries`. I simplified this code a bit, let's see if we need to work on this part some more.

ch1bo commented on hydra-node/src/Hydra/Chain/Blockfrost.hs:
```diff
@@ -65,3 +118,175 @@ mkTinyWallet tracer config = do
+  m ()
+blockfrostSubmissionClient prj tracer queue = bfClient
+ where
+  bfClient :: m ()
```
What's the point of this signature, lol

v0d1ch replied:
lol, probably copy/pasta from the chain observer.

ch1bo commented on hydra-node/src/Hydra/Chain/Blockfrost.hs:
```diff
@@ -65,3 +118,175 @@ mkTinyWallet tracer config = do
+  Blockfrost.Genesis{_genesisSlotLength, _genesisActiveSlotsCoefficient} <- Blockfrost.runBlockfrostM prj Blockfrost.getLedgerGenesis
+
+  Blockfrost.Block{_blockHash = (Blockfrost.BlockHash genesisBlockHash)} <-
+    Blockfrost.runBlockfrostM prj (Blockfrost.getBlock (Left 0))
```
Wait.. is this always starting at genesis??

Must: use the provided `chainPoint` to start following

v0d1ch replied:
No it uses `fromChainPoint` to pick the block hash. We could call blockfrost directly with the appropriate slot to get the block hash for it.

ch1bo reviewed (changes requested):
Could only spot one incorrectness (following from genesis if I read it correctly?)

Besides that.. lots of opportunity to make this cleaner architecturally and often shorter / clearer code.

v0d1ch committed:
PR review

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Rebase

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

locallycompact commented on hydra-node/golden/RunOptions.json:
```diff
@@ -8,33 +8,45 @@
+                    "18a22e4400bb478fec5dcad29b32aabd00b822aa721dccf8c9821354aabea153",
+                    "d0bafe586ebb838b9f8f8bb533892be0cb740a68d6f2b77b9f2d4f2c9ce08464",
+                    "b6fa856a9c21da5be600bda4da3b31dbbe213b0fe97f62340551e465fea3d963",
+                    "a726aedba74a0dc5f3ead4dc662fac9dfe1e566f88a9f7b9ebe7bf47e569df39"
```
What's happening here? Why so many script tx ids?

v0d1ch replied:
I guess the generator does it. I'll see if anything changed there.

v0d1ch replied:
Nothing changed but I added `reasonablySized` generator so we don't end up with a huge file.

ffakenz commented on hydra-cluster/src/Hydra/Cluster/Faucet.hs:
```diff
@@ -88,6 +97,55 @@ findFaucetUTxO RunningNode{networkId, nodeSocket} lovelace = do
+      let signedTx = signTx faucetSk tx
+      eResult <- Blockfrost.tryError $ Blockfrost.submitTransaction signedTx
+      case eResult of
+        Left err -> liftIO $ throwIO $ FaucetBlockfrostError{blockFrostError = show err}
```
i would use a SubmitTransactionException or similar instead

v0d1ch replied:
We have a dedicated type just for Blockfrost exceptions.

ffakenz commented on hydra-cluster/src/Hydra/Cluster/Faucet.hs:
```diff
@@ -88,6 +97,55 @@ findFaucetUTxO RunningNode{networkId, nodeSocket} lovelace = do
+        Left err -> liftIO $ throwIO $ FaucetBlockfrostError{blockFrostError = show err}
+        Right _ -> do
+          void $ Blockfrost.awaitUTxO networkId [changeAddress] (txId signedTx) 100
+          Blockfrost.awaitUTxO networkId [receivingAddress] (txId signedTx) 100
```
if awaitUTxO takes a list of addresses, why not awaiting for both at once and then return the appropriate?

v0d1ch replied:
true, I should do that.

v0d1ch replied:
ah no. If I do that then the produced UTxO will have entries for the faucet key as well. In the test we need to get only the relevant pub key output UTxO.

ffakenz commented on hydra-cluster/src/Hydra/Cluster/Faucet.hs:
```diff
@@ -88,6 +97,55 @@ findFaucetUTxO RunningNode{networkId, nodeSocket} lovelace = do
+      liftIO $
+        throwIO $
+          FaucetHasNotEnoughFunds{faucetUTxO}
+    pure $ maybe mempty (uncurry UTxO.singleton) foundUTxO
```
what is the difference between mempty and having not founds?

v0d1ch replied:
in this context it should be the same?

ffakenz commented on hydra-cluster/src/Hydra/Cluster/Util.hs:
```diff
@@ -66,7 +67,20 @@ chainConfigFor ::
+  ContestationPeriod ->
+  DepositPeriod ->
+  IO ChainConfig
+chainConfigFor' me targetDir socketOrProjectPath hydraScriptsTxId them contestationPeriod depositPeriod = do
```
why not taking chainBackendOptions as argument instead of socketOrProjectPath?

v0d1ch replied:
NetworkId is always the same so we differentiate only on either socket  or blockfrost project file.

ffakenz commented on hydra-cluster/test/Test/BlockfrostChainSpec.hs:
```diff
@@ -0,0 +1,182 @@
+
+spec :: Spec
+spec = around (showLogsOnFailure "BlockfrostChainSpec") $ do
+  it "can open, close & fanout a Head using Blockfrost" $ \tracer -> do
```
nice!

ffakenz commented on hydra-cluster/test/Test/BlockfrostChainSpec.hs:
```diff
@@ -0,0 +1,182 @@
+                | snapshotNumber == 1 -> Just contestationDeadline
+              _ -> Nothing
+
+          waitMatch aliceChain $ \case
+            Tick t _ | t > deadline -> Just ()
+            _ -> Nothing
```
very interesting way of awaiting for certain block

ffakenz commented on hydra-tx/src/Hydra/Tx/Abort.hs:
```diff
@@ -55,6 +55,7 @@ abortTx committedUTxO scriptRegistry vk (headInput, initialHeadOutput) headToken
             & addTxOuts reimbursedOutputs
             & burnTokens headTokenScript Burn headTokens
             & addTxExtraKeyWits [verificationKeyHash vk]
+            & setTxMetadata (TxMetadataInEra $ mkHydraHeadV1TxName "AbortTx")
```
thx! 👏

ffakenz commented on hydra-node/src/Hydra/Chain/Backend.hs:
```diff
@@ -0,0 +1,38 @@
+import Hydra.Chain.CardanoClient qualified as CardanoClient
+import Hydra.Tx (ScriptRegistry)
+
+class ChainBackend a where
```
in love with this type class! <3

locallycompact replied:
I'm not a fan of this kind of lawless overloading, but if we must do this then it should be a monad class. methods should not have constraints in the context.

An example of this kind of style is in pandoc:

https://hackage-content.haskell.org/package/pandoc-3.7.0.1/docs/Text-Pandoc-Class.html#t:PandocMonad

v0d1ch replied:
@locallycompact I lean towards your approach but even more I think handles would be appropriate considering the our coding style. Let's do the changes as a follow up if you agree since this PR is dragging for a long time.

ffakenz reviewed (approved):
nice job!

excited to see this in action already!

v0d1ch committed:
CI secret management

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Bump wait time for blockfrost query

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
No need to update networkId - this was causing failed tests

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Abstract common information into Hydra.Chain.Cardano

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Use reasonably sized RunOptions generator

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

noonio commented on CHANGELOG.md:
```diff
@@ -30,6 +30,8 @@ changes.
   - Change to the `ReqSn` message in the Hydra network protocol
   - Added `DepositExpired` for when a deposit was deemed expired.
 
+- Enable blockfrost integration for hydra-node.
```
Let's add an issue to add a documentation section about how to use it

noonio reviewed (approved):
🥳
